pennylane_lightning_kokkos/src/simulator/StateVectorKokkos.hpp

```diff
@@ -702,6 +702,187 @@ template <class Precision> class StateVectorKokkos {
+
+            for (size_t i = 0; i < indices.size(); i++) {
+                if (excludedIndex == indices[i])
+                    indices.erase(indices.begin() + i);
+            }
+        }
+        return indices;
+    }
+
+    auto generateBitPatterns(const std::vector<size_t> &qubitIndices,
+                             size_t num_qubits) -> std::vector<size_t> {
+
+        std::vector<size_t> indices;
+        indices.reserve(Util::exp2(qubitIndices.size()));
+        indices.emplace_back(0);
+
+        for (size_t index_it0 = 0; index_it0 < qubitIndices.size();
+             index_it0++) {
+            size_t index_it = qubitIndices.size() - 1 - index_it0;
+            const size_t value =
+                Util::maxDecimalForQubit(qubitIndices[index_it], num_qubits);
+
+            const size_t currentSize = indices.size();
+            for (size_t j = 0; j < currentSize; j++) {
+                indices.emplace_back(indices[j] + value);
+            }
+        }
+        return indices;
+    }
+
+    /**
+     * @brief Probabilities for a subset of the full system.
+     *
+     * @param wires Wires will restrict probabilities to a subset
+     * of the full system.
+     * @return Floating point std::vector with probabilities.
+     * The basis columns are rearranged according to wires.
+     */
+    std::vector<Precision> probs(const std::vector<size_t> &wires) {
+        // Determining index that would sort the vector.
+        // This information is needed later.
+        auto sorted_ind_wires = Util::sorting_indices(wires);
+        // Sorting wires.
+        std::vector<size_t> sorted_wires(wires.size());
+
+        for (size_t pos = 0; pos < wires.size(); pos++) {
+            sorted_wires[pos] = wires[sorted_ind_wires[pos]];
+        }
+
+        // Determining probabilities for the sorted wires.
+        const Kokkos::View<Kokkos::complex<Precision> *> arr_data = getData();
+
+        const size_t num_qubits = getNumQubits();
+
+        std::vector<size_t> all_indices =
+            generateBitPatterns(sorted_wires, num_qubits);
+
+        std::vector<size_t> all_offsets = generateBitPatterns(
+            getIndicesAfterExclusion(sorted_wires, num_qubits), num_qubits);
+
+        Kokkos::View<Precision *> d_probabilities("d_probabilities",
+                                                  all_indices.size());
+
+        Kokkos::View<size_t *> d_sorted_ind_wires("d_sorted_ind_wires",
+                                                  sorted_ind_wires.size());
+        Kokkos::View<size_t *> d_all_indices("d_all_indices",
+                                             all_indices.size());
+        Kokkos::View<size_t *> d_all_offsets("d_all_offsets",
+                                             all_offsets.size());
+
+        using UnmanagedSize_tHostView =
+            Kokkos::View<size_t *, Kokkos::HostSpace,
+                         Kokkos::MemoryTraits<Kokkos::Unmanaged>>;
+
+        Kokkos::deep_copy(
```

Comment:
In the case of checking beforehand if it's sorted, we will no longer need to copy these three d_ arrays

Reply:
yes, you're right for a full system, while for a sub-system, we have to copy all of them. will check it out.